### PR TITLE
#6326: Implement ConvTranspose2D

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**
### Ticket
#6326 

### Problem description
Add support for ConvTranpose2D. Needed for certain models and backpropagation.

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11597986321)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp [hunk 1]
```diff
@@ -499,6 +499,12 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_impl(
     uint32_t dilation_h = (uint32_t)sliding_window_config.dilation_hw.first;
     uint32_t dilation_w = (uint32_t)sliding_window_config.dilation_hw.second;
 
+    if(sliding_window_config.is_transpose)
+    {
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:503]: consistent formatting please

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 2]
```diff
@@ -32,10 +42,35 @@ Shape SlidingWindowConfig::get_output_shape() const {
...
 
+Shape SlidingWindowConfig::get_transposed_full_input_shape() const {
+    TT_ASSERT(is_transpose == true);
+    auto output_shape = get_output_shape();
+    uint32_t full_input_height = output_shape[1] + dilation_hw.first * (window_hw.first - 1);
+    uint32_t full_input_width = output_shape[2] + dilation_hw.second * (window_hw.second - 1);
+    return Shape( std::vector<uint32_t>{batch_size, full_input_height, full_input_width, 0});
+}
+
+std::array<uint32_pair_t, 2> SlidingWindowConfig::get_transposed_real_padding() const {
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:57]: What is "real" padding?

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 3]
```diff
@@ -32,10 +42,35 @@ Shape SlidingWindowConfig::get_output_shape() const {
         output_h = input_hw.first;
         output_w = input_hw.second;
     }
-    log_debug(tt::LogOp, "output_size: {} {} {}", batch_size, output_h, output_w);
+    log_debug(tt::LogOp, "SlidingWindowConfig::get_output_shape(): {} {} {}", batch_size, output_h, output_w);
     return Shape( std::vector<uint32_t>{batch_size, output_h, output_w, 0});
 }
 
+Shape SlidingWindowConfig::get_transposed_full_input_shape() const {
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:49]: What is "full" shape?

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 4]
```diff
@@ -32,10 +41,35 @@ Shape SlidingWindowConfig::get_output_shape() const {
...
+    TT_ASSERT(is_transpose == true);
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:49]: Please change to TT_FATAL

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 5]
```diff
@@ -32,10 +41,35 @@ Shape SlidingWindowConfig::get_output_shape() const {
...
+    auto output_shape = get_output_shape();
+    uint32_t full_input_height = output_shape[1] + dilation_hw.first * (window_hw.first - 1);
+    uint32_t full_input_width = output_shape[2] + dilation_hw.second * (window_hw.second - 1);
+    return Shape( std::vector<uint32_t>{batch_size, full_input_height, full_input_width, 0});
+}
+
+std::array<uint32_pair_t, 2> SlidingWindowConfig::get_transposed_real_padding() const {
+    TT_ASSERT(is_transpose == true);
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:57]: Please change to TT_FATAL

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+            case tt::ARCH::BLACKHOLE:
+                compute_kernel_config = BlackholeComputeKernelConfig(
+                    {.math_fidelity = conv_config.math_fidelity,
+                    .math_approx_mode = conv_config.math_approx_mode_enabled,
+                    .fp32_dest_acc_en = conv_config.fp32_dest_acc_enabled,
+                    .packer_l1_acc = conv_config.packer_l1_accum_enabled});
+                break;
+
+            default:
+                TT_ASSERT(false);
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:161]: Please change to TT_THROW

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+
+Tensor transform_weights_for_conv_transpose2d(const Tensor& conv_weight_tensor) {
+            switch (conv_weight_tensor.get_dtype()) {
+                case DataType::BFLOAT16:
+                    return _transform_weights_for_conv_transpose2d<::bfloat16>(conv_weight_tensor);
+                case DataType::FLOAT32:
+                    return _transform_weights_for_conv_transpose2d<float>(conv_weight_tensor);
+                case DataType::UINT32:
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:73]: do we support integer weights?

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            opt_conv_op_block_config,
+            conv_out_memory_config,
+            conv_config.dtype,
+            {batch_size, input_height, input_width, in_channels},
+            conv_config.input_channels_alignment == 16,
+            compute_kernel_config,
+            conv_config.enable_act_double_buffer,
+            conv_config.enable_split_reader,
+            conv_config.enable_subblock_padding);
+        // ttnn::operations::core::deallocate(halo_output);
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:254]: not deallocating based on config?

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp [hunk 9]
```diff
@@ -24,6 +25,14 @@ bool SlidingWindowConfig::has_parallel_config() const {
     * Calculate the window op output shape, excludes the channel dimension since this config is independent of the depth.
     */
 Shape SlidingWindowConfig::get_output_shape() const {
+    if(is_transpose) {
+        //Inverse of is_transpose == False
```
Comment (mywoodstock) [inline ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:29]: this comment needed?

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp (hunk 4 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:49]: Fixed.

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp (hunk 5 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:57]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 7 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:73]: I don't think the op supports it. But the existing weights preparation code [does](https://github.com/tenstorrent/tt-metal/blob/4e1fef96b1fc5543455911fef6fcf7664f04d9b3/ttnn/cpp/ttnn/tensor/tensor_utils.cpp#L204).

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 8 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:272]: It was disabled for debugging.

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:272]: Re-enabled now.

File: ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp (hunk 9 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:29]: I meant that it does the inverse calculation of the forward pass.

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/sliding_window/sliding_window.cpp:29]: Wrote a better comment.

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp (hunk 1 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp]: Fixed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 6 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Fixed.

File: tests/ttnn/unit_tests/operations/test_conv_transpose2d.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    input_width,
+    filter_height,
+    filter_width,
+    stride_h,
+    stride_w,
+    pad_h,
+    pad_w,
+    out_pad_h,
+    out_pad_w,
+    use_1d_systolic_array=True,
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply tests/ttnn/unit_tests/operations/test_conv_transpose2d.py:52]: Removed.

File: tests/ttnn/unit_tests/operations/test_conv_transpose2d.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        output_layout=ttnn.ROW_MAJOR_LAYOUT,
+    )
+    if config_override and "act_block_h" in config_override:
+        conv_config.act_block_h_override = config_override["act_block_h"]
+
+    if config_override and "act_block_w_div" in config_override:
+        conv_config.act_block_w_div = config_override["act_block_w_div"]
+
+    if config_override and "num_cores_nhw" in config_override:
+        if config_override["num_cores_nhw"] == 98:
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply tests/ttnn/unit_tests/operations/test_conv_transpose2d.py:137]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+            device,
+            input_tensor,
+            conv_config,
+            batch_size,
+            output_height,
+            output_width,
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride,
```
Comment (sankarmanoj-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:176]: @pavlejosipovic In ConvTranspose2d, the meaning of stride & padding are the inverse of Conv2d. Auto_shard would have to check if it's the tranposed op, and if so, account for it correctly.

**Review by mywoodstock — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+#include "ttnn/operations/sliding_window/halo/halo.hpp"
+#include "ttnn/operations/conv/conv2d/conv2d.hpp"
+
+namespace ttnn {
+
+namespace operations::conv {
+namespace conv_transpose2d {
+
+
+template <typename T>
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp:30]: Why is this exposed in a header?

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        uint32_t input_width,
+        std::array<uint32_t, 2> kernel_size,
+        std::array<uint32_t, 2> stride,
+        std::array<uint32_t, 2> padding,
+        std::array<uint32_t, 2> output_padding,
+        std::array<uint32_t, 2> dilation,
+        uint32_t groups,
+        std::optional<const ttnn::Tensor> bias_tensor = std::nullopt,
+        std::optional<const conv2d::Conv2dConfig> conv_config_ = std::nullopt){
+        return conv_transpose2d(input_tensor, weight_tensor, device, in_channels, out_channels, batch_size, input_height, input_width, kernel_size, stride, padding, output_padding, dilation, groups, bias_tensor, conv_config_);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp:69]: Why is impl in a header?

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+#include "ttnn/operations/matmul/device/matmul_op.hpp"
+#include "ttnn/types.hpp"
+#include "ttnn/tensor/tensor_utils.hpp"
+#include "tt_metal/impl/dispatch/command_queue.hpp"
+#include "tt_metal/common/math.hpp"
+#include "ttnn/operations/data_movement/pad/pad.hpp"
+#include "ttnn/operations/conv/conv2d/device/conv2d_op.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/operations/sliding_window/sliding_window.hpp"
+#include "ttnn/operations/sliding_window/halo/halo.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp:21]: Can we minimize amount of includes in the header by removing unnecessary things from it, like method implementstions?

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp]: Removed unnecessary includes.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d_pybind.cpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+namespace conv_transpose2d {
+
+void py_bind_conv_transpose2d(py::module& module) {
+
+    bind_registered_operation(
+        module,
+        ttnn::conv_transpose2d,
+        R"doc(
+        Conv Tranpose 2D
+        +-------------------+-------------------------------+---------------+-------------+----------+
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d_pybind.cpp]: Updated doc string based on format used for matmul.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp (hunk 13 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp (hunk 14 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+    uint32_t input_height,
+    uint32_t input_width,
+    std::array<uint32_t, 2> kernel_size,
+    std::array<uint32_t, 2> stride,
+    std::array<uint32_t, 2> padding,
+    std::array<uint32_t, 2> output_padding,
+    std::array<uint32_t, 2> dilation,
+    uint32_t groups,
+    std::optional<const ttnn::Tensor> bias_tensor,
+    std::optional<const conv2d::Conv2dConfig> conv_config_);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:304]: Please remove template instantiation. It is not necessary.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+#include "ttnn/operations/conv/conv2d/conv2d.hpp"
+
+namespace ttnn {
+
+namespace operations::conv {
+namespace conv_transpose2d {
+
+
+struct ConvTranpose2dOperation{
+    static std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::Tensor>> invoke(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.hpp:15]: Please do this
```
using MeaningOfThisUint32 = uint32_t;
using MeaningOfThatUint32 = uint32_t;
using Result = std::tuple<ttnn::Tensor, MeaningOfThisUint32, MeaningOfThatUint32, ttnn::Tensor, std::optional<ttnn::Tensor>>;
```

Lets reduce some duplication and rely on this using across the code to reduce pain..

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+                compute_kernel_config = BlackholeComputeKernelConfig(
+                    {.math_fidelity = conv_config.math_fidelity,
+                    .math_approx_mode = conv_config.math_approx_mode_enabled,
+                    .fp32_dest_acc_en = conv_config.fp32_dest_acc_enabled,
+                    .packer_l1_acc = conv_config.packer_l1_accum_enabled});
+                break;
+
+            default:
+                TT_THROW("Invalid Device Arch, Got {}",device->arch());
+        }
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:162]: Extract into a free function?

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+            opt_conv_op_block_config,
+            conv_out_memory_config,
+            conv_config.dtype,
+            {batch_size, input_height, input_width, in_channels},
+            conv_config.input_channels_alignment == 16,
+            compute_kernel_config,
+            conv_config.enable_act_double_buffer,
+            conv_config.enable_split_reader,
+            conv_config.enable_subblock_padding);
+        ttnn::operations::core::deallocate(halo_output);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:266]: What's the point of this explicit deallocation if the scope ends in the next line? 
Should be totally fine for halo Tensor to get deallocated in its destructor.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 19 above)
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:162]: Extract into a free function (in an anonymous namespace). Something like this:

```suggestion
    DeviceComputeKernelConfig compute_kernel_config = get_compute_kernel_config(conv_config);
```

Proper anonymous namespace will have to look like this to work safely with unity builds
```
namespace {
namespace CMAKE_UNIQUE_NAMESPACE {
.... your free function
}
}
```

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+using namespace tt;
+namespace ttnn {
+namespace operations::conv {
+using sliding_window::SlidingWindowConfig;
+using sliding_window::ParallelConfig;
+
+namespace conv_transpose2d {
+
+template <typename T>
+Tensor _transform_weights_for_conv_transpose2d(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:19]: oh host?

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+Tensor transform_weights_for_conv_transpose2d(const Tensor& conv_weight_tensor) {
+            switch (conv_weight_tensor.get_dtype()) {
+                case DataType::BFLOAT16:
+                    return _transform_weights_for_conv_transpose2d<::bfloat16>(conv_weight_tensor);
+                case DataType::FLOAT32:
+                    return _transform_weights_for_conv_transpose2d<float>(conv_weight_tensor);
+                case DataType::UINT32:
+                    return _transform_weights_for_conv_transpose2d<uint32_t>(conv_weight_tensor);
+                default: TT_THROW("Unsupported data type for transform_weights_for_conv_transpose2d",conv_weight_tensor.get_dtype());
+            }
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:76]: fix format 
```suggestion
    switch (conv_weight_tensor.get_dtype()) {
        case DataType::BFLOAT16:
            return _transform_weights_for_conv_transpose2d<::bfloat16>(conv_weight_tensor);
        case DataType::FLOAT32:
            return _transform_weights_for_conv_transpose2d<float>(conv_weight_tensor);
        case DataType::UINT32:
            return _transform_weights_for_conv_transpose2d<uint32_t>(conv_weight_tensor);
        default: TT_THROW("Unsupported data type for transform_weights_for_conv_transpose2d",conv_weight_tensor.get_dtype());
    }
```

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+                auto out_buffer_kw_index = kernel_width - in_kernel_width_index - 1;
+                for(auto in_channels_index = 0; in_channels_index < in_channels; in_channels_index++)   {
+                    for(auto out_channels_index = 0; out_channels_index < out_channels; out_channels_index++)   {
+                        auto in_idx = in_channels_index * kernel_height * kernel_width * out_channels + out_channels_index * kernel_height * kernel_width + in_kernel_height_index * kernel_width + in_kernel_width_index;
+                        auto out_idx = out_channels_index * in_channels * kernel_height * kernel_width + in_channels_index * kernel_height * kernel_width + out_buffer_kh_index * kernel_width + out_buffer_kw_index;
+                        output_buffer[out_idx] = input_buffer[in_idx];
+                    }
+                }
+            }
+        }
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:46]: Maybe smth like this. It avoids redundant multiplication and will probably improve data locality.

```
for (auto out_channels_index = 0; out_channels_index < out_channels; out_channels_index++) {
    auto out_base_idx = out_channels_index * in_channels * kernel_height * kernel_width;
    for (auto in_channels_index = 0; in_channels_index < in_channels; in_channels_index++) {
        auto in_base_idx = in_channels_index * kernel_height * kernel_width * out_channels + out_channels_index * kernel_height * kernel_width;
        for (auto in_kernel_height_index = 0; in_kernel_height_index < kernel_height; in_kernel_height_index++) {
            auto out_buffer_kh_index = kernel_height - in_kernel_height_index - 1;
            auto in_height_offset = in_kernel_height_index * kernel_width;
            auto out_height_offset = out_buffer_kh_index * kernel_width;
            for (auto in_kernel_width_index = 0; in_kernel_width_index < kernel_width; in_kernel_width_index++) {
                auto out_buffer_kw_index = kernel_width - in_kernel_width_index - 1;
                
                auto in_idx = in_base_idx + in_height_offset + in_kernel_width_index;
                auto out_idx = out_base_idx + out_height_offset + out_buffer_kw_index;

                output_buffer[out_idx] = input_buffer[in_idx];
            }
        }
    }
}
```

File: tests/ttnn/unit_tests/operations/test_conv_transpose2d.py [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+    #     torch.tensor(range(input_height * input_width)).reshape([1, 1, input_height, input_width]).float()
+    # )
+    torch_input_tensor_nchw = torch_input_tensor_nchw.broadcast_to(conv_input_shape).float()
+
+    torch_input_tensor = torch.permute(torch_input_tensor_nchw, (0, 2, 3, 1))
+    # torch_weight_tensor = (
+    #     torch.randn((1, output_channels, 1, 1), dtype=torch.bfloat16).broadcast_to(conv_weight_shape).float()
+    # )
+    torch_weight_tensor = torch.randn(conv_weight_shape, dtype=torch.bfloat16).float()
+    # torch_weight_tensor = torch.ones(conv_weight_shape, dtype=torch.bfloat16).float()
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_conv_transpose2d.py:84]: remove commented code please

File: ttnn/cpp/pybind11/operations/__init__.hpp [hunk 25]
```diff
@@ -120,6 +121,9 @@ void py_module(py::module& module) {
     auto m_conv2d = module.def_submodule("conv2d", "conv2d operation");
     conv::conv2d::py_bind_conv2d(m_conv2d);
 
+    auto m_conv_transpose2d = module.def_submodule("conv_transpose2d", "conv_transpose2d operation");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/pybind11/operations/__init__.hpp:124]: @mywoodstock , can you please revisit this part? I think there should be one module for convs and this should just be an op in that module.

Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp:124]: @sankarmanoj-tt please consider binding into the same module as defined on line 121. just call your registration into the same module.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 19 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:162]: [This PR ](https://github.com/tenstorrent/tt-metal/pull/14210) removes the compute kernel configs from Conv2dConfig. When those changes are included, this code will be removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 21 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:19]: Currently the weights preparation for Conv2d happens on host. It's a one-time operation, so made sense to implement on the host initially. But this is slow when loading large networks.  This must be ported to the device. 
Relevant issue : https://github.com/tenstorrent/tt-metal/issues/14080

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 17 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 20 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Removed.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 22 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Fixed.

File: tests/ttnn/unit_tests/operations/test_conv_transpose2d.py (hunk 24 above)
Comment (sankarmanoj-tt) [reply tests/ttnn/unit_tests/operations/test_conv_transpose2d.py]: Removed.

File: ttnn/cpp/pybind11/operations/__init__.hpp (hunk 25 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp]: Changed to a single py module.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 23 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Done.

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**